nox/_parametrize.py

```diff
@@ -32,70 +82,83 @@ def parametrize_decorator(arg_names, arg_values_list):
             argument names were specified, this must be a list of N-tuples,
             where each tuple-element specifies a value for its respective
             argument name, for example ``[(1, 'a'), (2, 'b')]``.
+        ids ([Sequence[str]): Optional sequence of test ids to use for the
+            parametrized arguments.
     """
 
-    # Allow args to be specified as any of 'arg', 'arg,arg2' or ('arg', 'arg2')
+    # Allow args names to be specified as any of 'arg', 'arg,arg2' or ('arg', 'arg2')
     if not isinstance(arg_names, (list, tuple)):
         arg_names = list(filter(None, [arg.strip() for arg in arg_names.split(",")]))
 
     # If there's only one arg_name, arg_values_list should be a single item
     # or list. Transform it so it'll work with the combine step.
     if len(arg_names) == 1:
         # In this case, the arg_values_list can also just be a single item.
-        if not isinstance(arg_values_list, (list, tuple)):
+        if isinstance(arg_values_list, tuple):
+            # Must be mutable for the transformation steps
+            arg_values_list = list(arg_values_list)
+        if not isinstance(arg_values_list, list):
             arg_values_list = [arg_values_list]
-        arg_values_list = [[value] for value in arg_values_list]
 
-    # Combine arg names and values into a list of dictionaries. These are
-    # 'call specs' that will be used to generate calls.
-    # [{arg: value1}, {arg: value2}, ...]
-    call_specs = []
-    for arg_values in arg_values_list:
-        call_spec = dict(zip(arg_names, arg_values))
-        call_specs.append(call_spec)
+        for n, value in enumerate(arg_values_list):
+            if not isinstance(value, Param):
+                arg_values_list[n] = [value]
+
+    # if ids aren't specified at all, make them an empty list for zip.
+    if not ids:
+        ids = []
+
+    # Generate params for each item in the param_args_values list.
+    param_specs = []
+    for param_arg_values, param_id in itertools.zip_longest(arg_values_list, ids):
+        if isinstance(param_arg_values, Param):
+            param_spec = param_arg_values
+            param_spec.arg_names = tuple(arg_names)
+        else:
+            param_spec = Param(*param_arg_values, arg_names=arg_names, id=param_id)
+
+        param_specs.append(param_spec)
 
     def inner(f):
-        previous_call_specs = getattr(f, "parametrize", None)
-        new_call_specs = update_call_specs(previous_call_specs, call_specs)
-        setattr(f, "parametrize", new_call_specs)
+        previous_param_specs = getattr(f, "parametrize", None)
```

Comment:
I had to do a double-take that this was a closure. May be less magical doing a top-level

```python
def _inner(param_specs, f):
    previous_param_specs = getattr(f, "parametrize", None)
    new_param_specs = update_param_specs(previous_param_specs, param_specs)
    setattr(f, "parametrize", new_param_specs)
    return f
```

and then here just returning

```python
return functools.partial(_inner, param_specs)
```

Reply:
Done.